### ISSUE TITLE
implement intptrcast model

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/miri#224]
Since right now miri's pointer comparison bails out on `&x as *const T < &y as *const T` but allows `&x as *const T == &y as *const T`, I suggest to also allow `<`. This will not impact ctfe-mode, since ctfe mode disallows any binary ops on pointers into different allocs, even `==`.

It's deterministic between multiple miri runs, doesn't leak `HashMap` ordering and only depends on optimizations and codegen from rustc (but that's already true for `==`).

Steps to do:
* [x] Implement force_bits, force_int and basic address assignment.
* [x] force_bits when doing primitive operations.
* [x] force_bits when doing casts.
* [x] force_ptr when creating a non-ZST reference. (Ideally through `check_ptr_access`? Redundant with validity checks but whatever.) We may need a `force_ptr_place` method or so; the freeze-sensitive visitor also needs that.
* [x] Make force_bits infallible (in Miri, not CTFE) -- see https://github.com/rust-lang/miri/issues/841.
* [ ] ~~force_bits when doing partial loads of a pointer -- and really everywhere we need a ptr, once it is total. (For now let's not do this, also see https://github.com/rust-lang/rust/issues/69488 for a similar problem around partially initialized data.)~~
* [x] Exploit "real" pointer alignment for alignment checks, but warn when we do. (But also see https://github.com/rust-lang/rust/issues/62420.)
* [x] Ptr-to-int of a dangling pointer (possible in safe code).
* [x] Fix the FIXMEs added in https://github.com/rust-lang/miri/pull/803.
* [x] Audit all remaining uses of to_bits (this includes the `to_<int type>` methods...), to_ptr and to_bits_or_ptr, all uses of is_ptr and is_bits, and also all uses of `force_ptr` and `force_mplace_ptr` (also `mplace_access_checked`, `check_mplace_access`) outside of `check_ptr_access`. Maybe replace to_{bits,ptr} by assert_{bits,ptr}?

**Comment by RalfJung** [comment]
My plan was to go even further and essentially implement the [intptrcast-model](https://github.com/nikomatsakis/rust-memory-model/issues/30). That would immediately make us support all pointer arithmetic.

Essentially, I think at some point miri should (in non-CTFE mode!) guarantee that safe Rust code can not error out (if the unsafe libraries it uses don't contain bugs). So hashing raw pointers and printing them and stuff like that all have to work.

**Comment by oli-obk** [comment]
Sounds reasonable to me. Once a pointer has been observed and thus it's integral representation is available, there's no way in safe code to go from that pointer back to the allocation, so we don't even need to support that direction, right? All we need to ensure is hat the integral representation doesn't change when obtaining it multiple times from a real pointer

We'll also need to generate an integer address when offsetting pointers due to overflow concerns. Overflowing exposes the address essentially.

**Comment by RalfJung** [comment]
> We'll also need to generate an integer address when offsetting pointers due to overflow concerns. Overflowing exposes the address essentially.

That's only true when you do the offsetting after casting to integer though, is it? The offset operations on pointers don't let safe code observe whether an overflow actually happened. Really, we could just generate an integer address immediately on allocation.

Doing lookup on a `Pointer` is still more efficient, so that may be a reason to convert back.

**Comment by oli-obk** [comment]
Well... You can do `wrapping_offset` (safe and stable method), and then do a `<=` comparison between the resulting pointer and the new one.

```rust
let x = 5;
for i in 1.. {
    let ptr = &x as *const i32;
    let ptr2 = ptr.wrapping_offset(i);
    if ptr2 <= ptr {
        println!("Address of `x` is {}", std::usize::max_value() - i);
    }
}
```

> Really, we could just generate an integer address immediately on allocation.

We'd run out of integer addresses eventually (and quickly on 32bit emulation) if we don't reuse integer addresses and create a unique integer address for every allocation.

**Comment by RalfJung** [comment]
Heh, nice catch.

> We'd run out of integer addresses eventually (and quickly on 32bit emulation) if we don't reuse integer addresses and create a unique integer address for every allocation.


"Quickly"? 2^32 is still a large number, I feel.^^
But anyway, I guess there's no problem with lazy allocation. That also gives us a nice single point of control for rejecting everything that could potentially leak the base address when in CTFE mode -- we just always make the allocation of an integer address fail in that mode.

**Comment by oli-obk** [comment]
The problem is, that once you leak the base address of an array, you also leak the address of the last element's last byte. So allocatingand deallocating 100MB 50 times would fill the address space unless we allow reusing it.

**Comment by RalfJung** [comment]
Actually, given what I wrote in <https://internals.rust-lang.org/t/types-as-contracts/5562/81>, I think it would make sense for miri to both support pointer comparisons on "abstract" pointers and only provide intptrcast optionally. Without intptrcast, there will always be safe code that miri cannot run (like printing the value of a pointer), but currently there still seems to be value in also having a mode that keeps pointers and integers more separate.

**Comment by Amanieu** [comment]
This code currently fails on miri:
```rust
fn main() {
    let x: Vec<u32> = Vec::new();
    &x == &[];
}
```
```
error[E0080]: constant evaluation error: attempted to do invalid arithmetic on pointers that would leak base addresses, e.g., comparing pointers into different allocations
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:5024:12
     |
5024 |         if self.as_ptr() == other.as_ptr() {
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempted to do invalid arithmetic on pointers that would leak base addresses, e.g., comparing pointers into different allocations
     |
     = note: inside call to `<[A] as core::slice::SlicePartialEq<A>><u32>::equal` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:4962:9
     = note: inside call to `core::slice::<impl std::cmp::PartialEq<[B]> for [A]><u32, u32>::eq` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/liballoc/vec.rs:2069:50
     = note: inside call to `<std::vec::Vec<A> as std::cmp::PartialEq<[B; _]>><u32, u32>::eq` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/cmp.rs:1009:48
note: inside call to `std::cmp::impls::<impl std::cmp::PartialEq<&'b B> for &'a A><std::vec::Vec<u32>, [u32; 0]>::eq` at src/main.rs:3:5
    --> src/main.rs:3:5
     |
3    |     &x == &[];
     |     ^^^^^^^^^
     = note: inside call to `main` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:64:34
```

**Comment by RalfJung** [comment]
Blocked on https://github.com/rust-lang/miri/issues/653

**Comment by oli-obk** [comment]
cc @christianpoveda

Basic (meaning probably incomplete) implementation instructions:

1. add a `fn int_ptr_to_ptr(&self, ptr: Scalar) -> InterpResult<'tcx, Pointer>` method to `Machine` (in rustc). This function should have a default impl that matches on `Scalar`, returns the `Pointer` if it's the `Ptr` variant and returns the `ReadBytesAsPointer` error variant otherwise
2. find all current sites where `ReadBytesAsPointer` is being emitted and call `M::int_ptr_to_ptr(self, scalar)` at said sites instead of emitting the error.
3. Do the same trick for `ReadPointerAsBytes` (but name the newly added method `ptr_to_int_ptr(&self, ptr: Pointer) -> InterpResult<'tcx, Scalar>` and fully ignore the argument and just emit the error variant)
4. Up to now, there should be zero change in behaviour by const eval or miri
5. go back to miri, all further changes should be doable completely within miri
6. add a `deterministc` boolean field similar to https://github.com/rust-lang/miri/blob/4e329ebab7bb313c66bc09e728d875352c5a10f9/src/bin/miri.rs#L143
7. uh... wait for oli to continue these instructions tomorrow

**Comment by christianpoveda** [comment]
Ok I'll start working on this :D

**Comment by christianpoveda** [comment]
Should I wait for https://github.com/rust-lang/rust/pull/59654?

**Comment by oli-obk** [comment]
nah. whoever comes last needs to rebase ;)

**Comment by RalfJung** [comment]
I think `int_to_ptr` and `ptr_to_int` would be better names? This is talking about *values*, not *types* -- so mentioning "int ptrs" and "ptr ints" seems just confusing.

> add a deterministc boolean field similar to 

IMO this could just use the `seed` flag? RNGs and ptr-int-casts are just two sources of non-determinism.

**Comment by oli-obk** [comment]
I modified the above comment to reflect this

**Comment by christianpoveda** [comment]
I'm on step 1 and got [errors](https://gist.github.com/christianpoveda/2e29ee2dd51e91eb6585a1fe7a998e84/raw/0ff048b3ca55a27c3d5c2e9b81f3afcc049c28fe/gistfile1.txt) when running tests, the linker is failing and I'm not even sure if its my fault, these are the changes: https://github.com/christianpoveda/rust/commit/1458fa3ca85e3572c565fe7e17a9e5edcc7b0da3

**Comment by RalfJung** [comment]
I never run `run-make-fulldeps` tests...^^

What is your `x.py` invocation?

**Comment by christianpoveda** [comment]
`. /x.py test`

**Comment by RalfJung** [comment]
When just working on the Miri engine, I'd recommend `./x.py test --stage 1 src/test/{ui,run-pass}`. That's much faster and still gives you very good test coverage.

The error you are running into looks like a bug, might be worth reporting if you can reproduce it on unchanged master.

**Comment by christianpoveda** [comment]
Oh great! I'll check on master to see what happens. I will continue and hope for the best during CI hehehe

**Comment by christianpoveda** [comment]
I advanced to step 2, i checked for all the usages of `ReadBytesAsPointer` and found these:
```
src/librustc_mir/transform/const_prop.rs
187:                    | ReadBytesAsPointer

src/librustc/mir/interpret/error.rs
255:    ReadBytesAsPointer,
349:            ReadBytesAsPointer =>

src/librustc/mir/interpret/value.rs
313:            Scalar::Bits { .. } => err!(ReadBytesAsPointer),

src/tools/miri/src/fn_call.rs
614:                    Scalar::Bits { .. } => return err!(ReadBytesAsPointer),
```
I'm not sure if `int_to_ptr` is a proper replacement for any of those places.

**Comment by RalfJung** [comment]
Some of those are just *catching* `ReadBytesAsPointer`, you don't need to change that.

The one in `value.rs` is the "original source" of them. But you won't be able to call `int_to_ptr` here.  I think you should create a helper method in `librustc_mir/interpret` (maybe in `eval_context.rs`) that is used instead of `to_ptr` in every place where we do want this cast to occur. The place you found in Miri needs special treatment for NULL, you can ignore that for now (but add a FIXME).

**Comment by christianpoveda** [comment]
Ok, I've added the helper method into `InterpretCx`, on which places do we want this cast?

**Comment by RalfJung** [comment]
Well, many of the places where we use `to_bits`/`to_ptr`. I'm afraid figuring out for which exactly to do what is a lot of work, and the reason why I have not implemented this myself yet.

Given that one can transmute between integers and pointers, and that not all valid pointers are even representable as `Scalar::Ptr`, I think it makes sense to do the conversion as lazy as possible.  So the two places where our hands are forced would be:

* in `operator.rs`, when performing an integer operation, we should to `ptr_to_int` instead of dispatching to `ptr_op`.
* whenever we access memory, *after* determining that this is not a ZST, we should do `int_to_ptr`.

**Comment by christianpoveda** [comment]
Ok, I'll start digging then :D

**Comment by christianpoveda** [comment]
This is an small update:

- The `int_to_ptr` method has been added and it is replacing some of the usages of `to_ptr` as discussed with @RalfJung.
- The `ptr_to_int` method has been added and I will start checking today where it can be used.

After that I think that this part is ready for PR, the other changes must be done in miri AFAIK. Is that ok?

**Comment by christianpoveda** [comment]
Ok I might have a problem, both `int_to_ptr` and `ptr_to_int` are static on my implementation, and of course they shoudn't. However, I don't know how to use these methods to replace usages of `to_ptr`/`to_bits`. For example, [here](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/memory.rs#L632) I cannot find a way to recover the `Machine` or the `InterpretCx` to call the methods accordingly.

**Comment by RalfJung** [comment]
The machine hooks should likely just take a `MemoryExtra`, not `self`.

**Comment by christianpoveda** [comment]
I've changed the `Machine` hooks and now both return an error. I've also moved the `force` methods to `Memory` and forwarded them in `InterpretCx`.

However, I've been replacing `to_ptr` by `force_ptr` in the places we discussed with @RalfJung and most of the time it causes UB or ends up with the following error when compiling `rustc`:
```
 a raw memory access tried to access part of a pointer value as raw bytes
```

Could you take a look? https://github.com/rust-lang/rust/compare/master...christianpoveda:intptrcast-model

**Comment by christianpoveda** [comment]
I replaced some of the `to_ptr` uses by the new `force_ptr` method. I am trying to do the same with `force_bits` in https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/operator.rs#L317

**Comment by RalfJung** [comment]
That looks odd, there shouldn't be two calls to `binary_int_op`. The entire point was to keep the code paths nice and uniform.

But I think @oli-obk is right, this is not very nice either. Let's move this refactoring to later, doesn't have to happen now. For now, keep the dispatching the way it was.

**Comment by oli-obk** [comment]
Some discussion between eddyb and me about miri and threads came to the conclusion that for certain kind of checks it would be helpful to only have the intptrcast part where you can go from ptr to int, but not the other way. I believe that would even cover most real use cases, as I've only seen the other direction in artificial examples so far

**Comment by RalfJung** [comment]
I don't understand what you even mean by that. Could you give some concrete examples of programs we should not support and why?

Also see [this Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/Miri.3A.20supporting.20more.20code.20vs.20detecting.20more.20bugs).

**Comment by oli-obk** [comment]
well, basically I want to forbid (on 64 bit)

```rust
union Foo {
    a: (u32, u32),
    b: usize,
}
let (x, y) = Foo { b: &1 as *const i32 as usize }.a;
let ptr = ((x as usize) << 32) | (y as usize);
let ptr = *(ptr as *const i32);
```

or any other bit modifying operations beyond those below the alignment (so those that will happen on the integer representation instead of on the abstract representation).

You can do these reading operations, so you can print all bits of a pointer, I just want to forbid `int2ptr` everywhere.

**Comment by RalfJung** [comment]
I'd rather get rid of all the code that performs arithmetic on the `Pointer` representation. That code is a mess.

Why don't you want to allow that code? I *do* want to forbid [this code](https://github.com/rust-lang/miri/blob/288ffa638fc0dd1ef2556e7cc5209f237d45efa8/tests/compile-fail/intptrcast_alignment_check.rs) which is why I added a test for this, but yours seems fine to me.

**Comment by oli-obk** [comment]
> Why don't you want to allow that code?

Essentially to be able to do full thread separation via our symbolic pointers. I want to be able to leverage the knowledge about what is accessible where. If we have integer pointers, that becomes impossible.

**Comment by RalfJung** [comment]
I'm afraid you will have to tell me more, I still don't know what this is all about.^^

Can you open a new issue for this specific restriction, and describe in more detail the why and how? My plan is to close this one once everything works.

I should probably put a checklist in the first comment here for what is still on @christianpoveda's TODO list. ;)

**Comment by christianpoveda** [comment]
Plz do

**Comment by RalfJung** [comment]
Done. :)

**Comment by RalfJung** [comment]
Turns out instead of calling `force_ptr` in `Ref`, we can have the invariant that places are actually "normalized" to `Pointer` on `Deref`, which happens in https://github.com/rust-lang/rust/pull/63075. So that checkbox is also taken care of.
